### ISSUE TITLE
New Calendar Control Design Proposal

### ISSUE THREAD

**Issue by robloo** [AvaloniaUI/Avalonia#5656]
# New Calendar Control Design Proposal

Originally taken from the discussion here: https://github.com/microsoft/microsoft-ui-xaml/issues/2255.

We need to improve editing of the selected date in the Calendar and by extension the CalendarDatePicker. These improvements are necessary to speed up editing by the user.

## Summary

The current Calendar is slow and somewhat unintuitive to use when editing a date far away from the currently selected one. It is deficient in the following areas:

  1. There is no visual indication that the title can be pressed (referring to the "month year" not the arrows)
  2. Pressing the title once goes to month view, twice goes to year view, and on older controls (WPF?) you could even press it a third time and go to a decade view I believe. This multiple pressing is not always ideal.
  3. The year, month, and day must all be set at once. **There is no mechanism to modify only the year or month of the currently selected date.** This means just changing the year requires re-selecting the month and day.

Each of these three points can be improved at once.

**Current Design**
![Untitled_1](https://user-images.githubusercontent.com/17993847/78815041-e1450280-798c-11ea-8ed8-38a9f835fb1e.png)

**Proposed Design**
Credit the illustrations goes to @mdtauk [comment](https://github.com/microsoft/microsoft-ui-xaml/issues/2255#issuecomment-611719568). (forgive me for putting this in the Avalonia repo!)

The top of the Calendar can now independently select day/month/year. Note that the control logic would also change and it will no longer be necessary to set the year, then always the month and day. Any component of the date will be modifiable and the other already selected components will persist.

![Untitled](https://user-images.githubusercontent.com/17993847/79055289-1eb5c400-7c09-11ea-97b0-a7459743caa9.png)

**Comment by robloo** [comment]
I should add a comment about the arrows that have been removed.

 - It's possible these aren't needed and the scroll wheel will be enough for end users. 
 - Adopt the concept from the DatePickerFlyoutPresenter and add overlay arrows when the mouse is over the calendar grid of numbers. 

![exaaaaaa](https://user-images.githubusercontent.com/17993847/111039315-5faaef80-83fb-11eb-99fc-7035dadc5fa0.gif)

If there are no objections, I'll take a look at making these changes myself.

**Comment by maxkatz6** [comment]
Am I right that WinUI hasn't updated CalendarView style yet with new Fluent styles? As I can see, there is only partial updates like background sizing and chevron-icons replacing.

**Comment by maxkatz6** [comment]
> and add overlay arrows when the mouse is over the calendar grid of numbers.

Should these overlay arrows be full-width?

**Comment by robloo** [comment]
> Am I right that WinUI hasn't updated CalendarView style yet with new Fluent styles?

Haven't checked. I'm not expecting any significant changes in this area though. If you are really worried about conflicts I can review the next time I build WinUI.

>>and add overlay arrows when the mouse is over the calendar grid of numbers.
>
>Should these overlay arrows be full-width?

If the arrows are required, yes, my thinking was arrows over the full width of the calendar (Sunday - Saturday). The illustration from the DatePickerFlyoutPresenter was only to communicate the idea for those not familiar -- not really to give an accurate representation of how it would look.

I'm still thinking that the arrows may not be required though. ComboBox doesn't use them and it's intuitive to the user they need to scroll to see more items. In touch scenarios it's even more intuitive that the calendar numbers can be interacted with directly.

**Comment by robloo** [comment]
>>Am I right that WinUI hasn't updated CalendarView style yet with new Fluent styles?
>
>Haven't checked. I'm not expecting any significant changes in this area though. If you are really worried about conflicts I can review the next time I build WinUI.

I stand corrected, there are several design updates planned more significant than I expected: https://github.com/microsoft/microsoft-ui-xaml/blob/5f8e2cd4c4d4e72f4589fe1fbf41c1d90be826db/specs/CalendarView/CalendarViewSpec1.md

**Comment by robloo** [comment]
Note that these modifications could actually allow removing DatePicker in the future in favor of CalendarDatePicker. CalendarDatePicker would then be just as fast to use as DatePicker; however, it has the added benefit of visually showing the calendar instead of large scrolling lists in DatePicker. This would be a much better user experience and eliminate one control to maintain in Avalonia.

**Comment by amwx** [comment]
I've been wanting to port the WinUI CalendarView for some time now (well reverse engineer it anyway), but in doing so adapt the design from mdtauk because I <i>really</i> like that over the original design. I was actually working on that back when I ported DatePicker/TimePicker, but I was far less experienced with Avalonia at the time and it was just bad.

IIRC, the WPF based Calendar control that exists currently is the slowest control to create in Avalonia and refactoring that to make it better has been on "the list" for a while but no one has taken it up.

**Comment by robloo** [comment]
@amwx Yea, mdtauk's design really turned the idea into something everyone can appreciate, and he pulled it all together quite nicely (though I take a little credit for the original concept and steering things :) ). I'm glad others like this idea!

I'm happy to support in any re-work of this control. It's poor quality code from my initial look. I can imagine it has performance too considering that the code looks pieced together instead of designed from the start. I decided to do some refactoring in CalendarDatePicker before I felt comfortable adding pseudoclasses to it. I don't know how Microsoft ever shipped this! I guess it has been so well tested by now there are enough patches it can hold water in production.

As always, I would wait for WinUI 3 to release before re-working these types of controls. It would benefit to have WPF/UWP source code visible at the same time to see what changes were made. As UWP was a newer re-write I expect it's a bit better in areas. Who knows when/if that is ever going to happen though.

**Comment by amwx** [comment]
> As always, I would wait for WinUI 3 to release before re-working these types of controls. It would benefit to have WPF/UWP source code visible at the same time to see what changes were made. As UWP was a newer re-write I expect it's a bit better in areas. Who knows when/if that is ever going to happen though.

I still might give it another shot, even if just in my own library (as more a POC). As far as I know, especially since it has different behavior from the WPF days, it was a complete rewrite.

It also looks like Uno got their hands on the actual source code (the MIT license header at the top of the file is the giveaway). The conflicting license info is confusing, but even if it can't be used, serves a good indication that it is entirely different: https://github.com/unoplatform/uno/blob/master/src/Uno.UI/UI/Xaml/Controls/CalendarView/CalendarViewDayItem.cs

**Comment by robloo** [comment]
> It also looks like Uno got their hands on the actual source code (the MIT license header at the top of the file is the giveaway)

Yes, Uno got early access to the Grid and Calendar source code in WinUI 3 and made their ports of those two controls using the actual WinUI3 code. However, that was a direct license between Uno and Microsoft (3rd parties not included) and Uno itself is Apache. So the code cannot be used here (The licensing is very suspect in this case). Note that Uno Platform also has full access to the entire WinUI3 code-base. A perk of having the right connections.

**Comment by amwx** [comment]
> > It also looks like Uno got their hands on the actual source code (the MIT license header at the top of the file is the giveaway)
> 
> Yes, Uno got early access to the Grid and Calendar source code in WinUI 3 and made their ports of those two controls using the actual WinUI3 code. However, that was a direct license between Uno and Microsoft (3rd parties not included) and Uno itself is Apache. So the code cannot be used here (The licensing is very suspect in this case). Note that Uno Platform also has full access to the entire WinUI3 code-base. A perk of having the right connections.

That's what I figured, and why I err on the side of caution with legal stuff. And like I said, its a good reference that its not the same as WPF. Now if only MS would actually release the WinUI3 source... But anyway, a redesign is needed. Actually, since I did the fluent styles port for Calendar, I remember there's a hard coded width in the Calendar to keep it the same size when switching to year view which I've never liked since its not extensible. But at the time I didn't want to touch the C# part of the control cause it is a large (and fragmented) control and like I said I wasn't familiar enough with Avalonia to change that at the time.

**Comment by robloo** [comment]
GTK has a good idea to solve the "arrows" problem. They have added left/right arrows beside both the month and year in the header. This allows for much faster navigation and could be integrated well with the new design proposal above. With this, I believe all functional and design issues can be resolved in this proposal -- nothing is stopping an implementation.

![calendar](https://user-images.githubusercontent.com/17993847/208251810-74e75129-fa53-48d9-832c-1f4868ce4a60.png)

https://docs.gtk.org/gtk4/class.Calendar.html

**Comment by rabbitism** [comment]
@robloo I think most web components have very good design like below. 
![image](https://user-images.githubusercontent.com/14807942/208284884-229963ba-891b-4ea8-ba36-3925b75ecdb3.png)

**Comment by robloo** [comment]
@rabbitism That design doesn't seem to support fast and separate editing of day, month, and year. You still have to scroll/click through pages to find the month with the date you want.

The original design above has three separate views to edit the date. And a novel way of accessing these views faster than the current control. The navigation arrows are supplemental and not a primary feature I would say. Design-wise the tab interface looks better to me as well.

**Comment by kvalitetskontrollMarkaryd** [comment]
I would like to see a "clear selection button" on the bottom of the fly out too. And maybe a "today button" too.

**Comment by maxkatz6** [comment]
In a context of DateOnly change, we might try to use polyfill with TypeForwardedTo on non-net6.0 builds. https://github.com/Tenacom/PolyKit/blob/main/src/PolyKit.Polyfills/System/DateOnly.cs (as an example)

Might be potentially problematic, though ecosystem was full of similar polyfills in the era of netstandard1.6/2.0 (i.e. IAsyncEnumerable and friends).
